### PR TITLE
Convert `@babel/plugin-transform-spread` to TypeScript

### PR DESCRIPTION
**Pull request by wafuwafu13**
| Q                        | A 
| ------------------------ | ---
| License                  | MIT

**Before**
Only indistinct type here.
https://github.com/babel/babel/blob/69ae2388f818e9f8dd1d0fd7527a7c7d39990eeb/packages/babel-plugin-transform-spread/src/index.js#L42

**After**
Convert to TypeScript.
Does not touch logics.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/13770"></a>

### CODE REVIEW

File: packages/babel-plugin-transform-spread/src/index.ts [hunk 1]
```diff
@@ -33,15 +44,41 @@ export default declare((api, options) => {
     return false;
   }
 
-  function push(_props, nodes) {
+  function push(
+    _props: Array<
+      | t.Expression
+      | t.SpreadElement
+      | t.JSXNamespacedName
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-spread/src/index.ts:51]: Q: When can this be a `JSXNamespacedName`?

File: packages/babel-plugin-transform-spread/src/index.ts [hunk 2]
```diff
@@ -33,15 +44,41 @@ export default declare((api, options) => {
...
+      | t.ArgumentPlaceholder
+    >,
+    nodes: Array<t.Expression>,
+  ): Array<
+    t.Expression | t.SpreadElement | t.JSXNamespacedName | t.ArgumentPlaceholder
+  > {
     if (!_props.length) return _props;
-    nodes.push(t.arrayExpression(_props));
+    nodes.push(
+      t.arrayExpression(_props as Array<t.Expression | t.SpreadElement>),
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-spread/src/index.ts:60]: If this only works when `_props` is `Array<t.Expression | t.SpreadElement>`, it should be reflected in the function signature.

File: packages/babel-plugin-transform-spread/src/index.ts (hunk 1 above)
Comment (wafuwafu13) [reply packages/babel-plugin-transform-spread/src/index.ts]: `CallExpression.arguments` and `NewExpression.arguments` type is `(t.Expression | t.SpreadElement | t.JSXNamespacedName | t.ArgumentPlaceholder)[]`, but this cannot be `t.JSXNamespacedName` nor `t.ArgumentPlaceholder` so I fixed. 

https://github.com/babel/babel/pull/13770/commits/fdab8e01226f52e6e99e36258015e59a2eb1cb38
>const args = node.arguments as Array<t.Expression | t.SpreadElement>;

**Review by nicolo-ribaudo — APPROVED** [review]
Thanks!

**Review by JLHwung — APPROVED** [review]
Thanks!

**Pull request merged**